### DISCUSSION TITLE
Nouislider doesn't react to changes in Alpinejs variable nor Livewire which is entangled with alpine

### DISCUSSION THREAD

**Discussion by marttheis** [alpinejs/alpine#4011] [1. Help]
Hello everyone,

I have the following x-data scope:

       ` <div   x-data="{ sliderStoplossPercentageValue: @entangle('StopLossPC')}" >`
Inside that scope there is a inputfield:

```
<input type="number" name="in_stopLossPercentage" id="in_stopLossPercentage" placeholder=""
          wire:model.live.debounce.500ms="StopLossPC"  
          x-model="sliderStoplossPercentageValue"
            />
```
Then comes the Nouislider DIV followed by the script:

 ```
<div   id="in_stopLossPercentageSlider" wire:ignore></div>

<script>
                     
var sliderStopLossPercentage = document.getElementById('in_stopLossPercentageSlider');
var sliderStopLossPercentageValue = document.getElementById('in_stopLossPercentage');

if (sliderStopLossPercentage.noUiSlider) {

            }
            else
            {
                noUiSlider.create(sliderStopLossPercentage, {
                    start: [7],
                    step: 0.5,
                    tooltips: true,
                    animate: true,
                    behaviour: 'none',
                    animationDuration: 1600,
                    connect: 'lower',
                    range: {
                                                'min': [0.1, 0.1],
                                                '60%': [10, 0.1],
                                                '70%': [25, 1],
                                                '80%': [50, 5],
                                                'max': [100]
                                            }
                    
                });
                sliderStopLossPercentage.noUiSlider.on('end', function (values, handle) {
                @this.StopLossPC=values[handle];
                console.log('slider sets the variables')
                });
                
            };
        

</script>
```
**Here is what works:** The slider sets the Alpine variable sliderStopLossPercentageValue and livewire changes accordingly due to being entangled. There are other inputfields as well and when their input changes the value of my input field above reflects that change based on backend calculations. This change also immediately changes the alpine variable. Everything is nicely synced when it comes to my Alpine and Livewire variables.

**What doesn't work:** I can't find a way to set the slider to the current value of the alpinevariable when the variable changes. I tried livewire.hook because alpine is entangled but it didn't work either. There must be a simple way of watching the alpine variable for changes and then set the slider accordingly. However I can't find it. Of course I also want the slider to change when I make a manual input in the input field. This is easy to do but fails to work for non manual changes.

The solution must be super simple because I have the variables with correct values right there next to the slider but simply can't connect both. Maybe alpine.effect? I tried but got errors. 

I am not an experienced coder at all. Help is much appreciated!

**Disclosure: I also asked this question on stackoverflow but got no reply yet.**

**Comment by ekwoka** [comment]
You need to look at the sliders API and see if it has an update options, (or just destroy and recreate) and register to do that with x-effect or Alpine.effect.

Alpine can't make other things register effects.

**Reply by marttheis** [reply]
I did some more tests and learned that x-model="sliderStoplossPercentageValue" in the input field is actually redundant because entanglement will eventually change that sliderStopLossPercentageValue as well, it just takes a little longer due to the backend calculations.

However! I also learned that I am basically unable to actually get the above alpinevariable correctly into the JS script. It is always null or undefined. 

The sliders have a on change, update or set methods which I could use if the value of the alpinevariable would be available within the script. 


I tried to move the x-data into JS as follows. I also removed the x-model="StopLossPercentageValue" from the input and added the script initalization to the slider div via x-init= "setslider()". I added a span with x-text to see if the entanglement actually updates the alpine Variable outside of the js script. That worked nicely as explained above. I also removed the `var sliderStopLossPercentageValue = document.getElementById('in_stopLossPercentage');` because it was redundant as well. But I am not sure.

Here is the code. Setting and getting the slider needs to be done within the setslider() function eventually.

```
<div x-data="test()">

        <input type="number" name="in_stopLossPercentage" id="in_stopLossPercentage" placeholder=""
          wire:model.live.debounce.500ms="StopLossPC" />

        <div  x-init= "setslider()" id="in_stopLossPercentageSlider" wire:ignore></div>


         <span x-text="StopLossPercentage"></span>


<script>

function test() {

return {

StopLossPercentageValue @entangle('StopLossPC'),

function setslider() {

console.log(this.StopLossPercentageValue}

}}}
</script>

</div>
```


The result is that when I trigger a backend calculkation so that StopLossPC changes it will take a seconds and the x-text also changes to that value due to being entangled. No x-model needed here. 

However the console returns Null or undefined for `console.log(this.StopLossPercentageValue}` from within the script.

So I can entangle StopLossPercentageValue with livewire in the script and then use it within alpine but not within javascript via` this.`

Is there a way to make StopLossPercentageValue available in the script? Once that works it should be simple to sync the slider to that value. 

The last option would be to create an event in livewire when StopLossPC is changing and then listen to that event and set the slider to the StopLossPC value.

However, I think it would be much nicer to sync it to the alpine variable.

How can I make StopLossPercentageValue availabe in the script?

best regards
Martin

**Reply by ekwoka** [reply]
is that the code you have in the component? because it's not valid javascript at all.

what is the initial state of the dom when it's sent?

**Reply by marttheis** [reply]
It's something I tried following this approach here: https://codewithhugo.com/alpinejs-x-data-fetching/

**Reply by marttheis** [reply]
I am not a coder so knowing what the initial state of the dom is is over my head unfortunately. But if you would point me in the right direction I will figure it out. 

everything works nicely except the fact that the slider don't automatically react when the alpine variable StopLossPercentageValue or the Livewire variable StopLossPC is changing. In addition I have the problem that I can't access the alpine variable StopLossPercentageValue inside the script tags. My second code example tried to solve this by putting the x-data in a function and accessing the alpine variable with `this.`  But that didn't work. 

There must be a way to have the those variables inside the script tags so that I can use them to set the slider.

**Reply by marttheis** [reply]
Thank you for your patience.
